### ISSUE TITLE
SC2234

### ISSUE THREAD

**Issue by accountnumber3** [koalaman/shellcheck#2169]
#### For bugs
- Rule Id: SC2234
- My shellcheck version: 0.7.1 as bundled with https://github.com/timonwong/vscode-shellcheck.git
- [ X ] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [ X ] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [ X ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash
functionName()([[ $(echo true) ]])

```

#### Here's what shellcheck currently says:
```
Line 2:
functionName()([[ $(true) ]])
              ^-- SC2234: Remove superfluous (..) around test command.
```

#### Here's what I wanted or expected to see:
Nothing, it's valid syntax. It's a simple function that just runs a single command and retains the return code. May not be the most optimized, but it's for compatibility with code that isn't mine.
Technically functions should be wrapped in brackets, but according to the stackoverflow post I saw last night, using parenthesis instead of brackets limits variable scope, and I don't want the output of the command to accidentally come with some extra variables.

**Comment by ingydotnet** [comment]
I just hit this. Here's a real world example:
```
on-mac() ( [[ $OSTYPE == darwin* ]] )
```

Also get the warning on these variants:
```
on-mac() ([[ $OSTYPE == darwin* ]])
on-mac() ( [[ $OSTYPE == darwin* ]]; )
on-mac() (
  [[ $OSTYPE == darwin* ]]
)
```

I prefer to always use parens over brackets for functions.
They limit more side effects than just variable state.
Bracket functions in my Bash code indicate that the function is purposefully setting some state for the caller.

**Comment by ingydotnet** [comment]
@koalaman As an aside here, shellcheck has usually concentrated on pointing out things that could be harmful, not things that are superfluous.

It makes me wonder if shellcheck should have options for what kind of things it will look for.

For instance the following double quotes are all superfluous:
```
foo="$baz"
foo="xxx${baz}yyy"
foo="$( ls -1 )"
[[ "$foo" == bar ]]
```

I've always wished there was another level for shellcheck to tell me that I'm doing it poorly. :)

**Comment by TinCanTech** [comment]
> I've always wished there was another level for shellcheck to tell me that I'm doing it poorly. :)

`  -S SEVERITY         --severity=SEVERITY        Minimum severity of errors to consider (error, warning, info, style)`

Would _the droid you're looking for_ be `style` ? :telescope: :milky_way:

**Comment by ingydotnet** [comment]
@TinCanTech I _knew_ I should have reviewed the `--help`!

That's the right track, but I think I need `-s nitpick` or somesuch.

The `style` severity doesn't catch my superfluous quotes.

So then I went down the rabbit hole of setting up my own style in a `.shellcheckrc`.

Unfortunately the things that I nitpick about aren't options yet.

For example: https://github.com/koalaman/shellcheck/wiki/SC2243 says this is problematic:
```
if [ "$(mycommand --myflags)" ]
then
  echo "True"
fi
```

I agree that it is, but I'd want:
```
if [[ $(mycommand --myflags) ]]; then
  echo True
fi
```

ie:
* Always `[[`
* Never `-n`
* No quote on var expansion in `[[`
* No quoting simple literal words
* Always `]]; then`

Might be time to put that rusty Haskell knowledge to work... :)

**Comment by TinCanTech** [comment]
I've never actually seen a `shellcheck` `style` comment myself, so maybe it is under used.

I mean, I can see the point you are making and I'm wondering if such a warning could be downgraded ?

It is a tricky question and `shellcheck` is like first line defence only..
`shellcheck` often interprets things in a manner which does not suit my style, sometimes I change my style and sometimes I just ignore `shellcheck`.. Hard to put a pin in that one.

FYI: I can never find the droids I am looking for ;-)

**Comment by grahamc** [comment]
I hit this as well in the following case:

```bash
is_integer() (
    [[ $1 =~ ^-?[0-9]+$ ]]
)
```

I worked around it by naming the argument:

```bash
is_integer() (
    arg=$1
    [[ $arg =~ ^-?[0-9]+$ ]]
)
```

**Comment by TinCanTech** [comment]
I notice that https://github.com/koalaman/shellcheck/wiki/SC2234 states:
- **Exceptions None**.

Perhaps, some of _these_ exceptions could be added to the wiki to lend some weight to the ideas here.

**Comment by ingydotnet** [comment]
Came to report the `foo() ( [[ $x ]] )` false positive I just hit only to find I've been here before!

Be well!

**Comment by koalaman** [comment]
All the examples here are true positives for the bad practices that ShellCheck aims to discourage.

I do agree that 1. the message made it seem like a purely stylistic issue, 2. the wiki exception part didn't help, and 3. the wrong advice was emitted for the right reason for function bodies since `f() [ x ]` is not valid, and ShellCheck purposefully doesn't allow `f() [[ x ]]` to improve diagnostics. Additionally, while it was only ever meant to trigger on side effect free code, it did not properly handle the case of mutating ${} and $(()) expansions like `f() ( [[ $((i++)) -eq 0 ]] )`. All these issues have now been fixed.

The original example now results in a better (and intentional) "SC2235: Use { ..; } instead of (..) to avoid subshell overhead", and only on side effect free `test` expressions.

**Comment by ingydotnet** [comment]
Thank you!

**Comment by ingydotnet** [comment]
I just realized `f() [[ ... ]]` is valid Bash. Good to see you already know it.
